design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,277 @@
+value will be verified against the key schema but will not populate the corresponding field in the 
+value. Populating columns with `ROWTIME` will not be supported.
+
+### Serialization Support
+
+JSON, AVRO and DELIMITED value formats will all be supported, the latter of which will not accept
+complex types. Since we know the schema of the inserts, we can initially load the values into a 
+connect data type. We already have the proper serializers to convert from connect to each of the
+supported serialization formats.
+
+## Key Handling
+
+The underlying data and the way it is represented in KSQL sometimes differ, and this `INSERT INTO`
+functionality needs to bridge that gap. For example take the following KSQL statement:
+```sql
+CREATE TABLE bar (id VARCHAR, foo INT) WITH (key="id")
+```
+
+In this case, we expect kafka records that resemble 
+```json
+{"key": "bar", "value": {"id": "bar", "foo": 123}}
+```
+
+If this is not the case (e.g. the value of `"key"` and of `"value"->"id"` do not match), then the
+KSQL behavior is not well defined. To all for all levels of flexibility, `INSERT INTO` should allow
+for all cases that kafka supports. Therefore, `ROWKEY` will always be available in the list of
+column names as the first column. The following examples would all be possible: 
+
+```sql
+INSERT INTO bar (ROWKEY, ID, FOO) VALUES ("k", null, 123);
+INSERT INTO bar (ROWKEY, ID, FOO) VALUES ("k", "k", 456);
+INSERT INTO bar (ROWKEY, ID, FOO) VALUES (null, "k", 789);
+INSERT INTO bar (ROWKEY, ID, FOO) VALUES ("k", "x", 1011); 
+INSERT INTO bar (ROWKEY, ID, FOO) VALUES ("k", null, null); 
+```
+
+If `ROWKEY` is _not_ specified but the key field is specified (as defined by the corresponding 
+`CREATE TABLE` statement), then `ROWKEY` can be inferred, meaning the following statements are
+equivalent:
+
+```sql 
+INSERT INTO bar (ID, FOO) VALUES ("k", 123)
+INSERT INTO bar (ROWKEY, ID, FOO) VALUES ("k", "k", 123)
+```
+
+## Test plan
+
+This functionality is expected to work both only in Interactive mode, and these tests must succeed:
+
+* Test that values can be serialized into JSON/AVRO/DELIMITED
+* Test that schemas are properly verified
+* Test `NULL` and table tombstones
+* Test `CREATE STREAM/TABLE` with topics that don't exist
+* Test producing data in topics that already exist
+* Test producing data to topics that are not associated with a KSQL source
+
+## Documentation Updates
+
+* The `CREATE TABLE` and `CREATE STREAM` syntax references will have an updated table properties
+section that includes the following rows:
+> ```rst
+> +=========================+======================================================================================================+
+> | KAFKA_TOPIC (required)  | The name of the Kafka topic that backs this source. The topic must either already exist in Kafka, or |
+> |                         | PARTITIONS and REPLICAS must be specified to create the topic. Command will fail if the topic exists |
```

Comment:
> PARTITIONS and REPLICAS must be specified to create the topic

Can we not just use some defaults? Seems laborious to specify them each time (given that the topic probably won't exist in many use cases)

Reply:
The current defaults for `CSAS` and `CTAS` since 5.2 take these properties from the source topic (we deprecated the default Partitions/Replicas because they were causing some problems - cc @hjafarpour). Since it's always easy to add defaults later but impossible to remove them, I think I'd err toward the verbose side and improve it later if it is a serious concern.

Reply:
I like the fact that if `PARTITIONS` and `REPLICAS` are not supplied then the command will fail with a suitable error message.  This gives users the ability to have it fail if the topic does not exist, while also have the power to auto-create when they need to.

Best of both worlds :D